### ISSUE TITLE
🐛 Need more flexible output/source directory ingress

### ISSUE THREAD

**Issue by sgiavasis** [FCP-INDI/C-PAC#1723]
### Describe the bug

C-PAC v1.8+ allows the ingress of existing pipeline output BIDS directories, either by providing:

- an [output BIDS directory](https://github.com/FCP-INDI/C-PAC/blob/80424468c7f4e59c102f446b05d4154ec1cd4b2d/dev/docker_data/default_pipeline.yml#L17-L21) that already has data in it, and C-PAC then processes what's left and writes the results back out into that output directory
- a [source BIDS directory](https://github.com/FCP-INDI/C-PAC/blob/80424468c7f4e59c102f446b05d4154ec1cd4b2d/dev/docker_data/default_pipeline.yml#L23-L30), which accepts another pipeline output directory, but intended for C-PAC to pull from this and then write new results to [the output directory](https://github.com/FCP-INDI/C-PAC/blob/80424468c7f4e59c102f446b05d4154ec1cd4b2d/dev/docker_data/default_pipeline.yml#L15-L21), presumably a new/different one

This is [the function](https://github.com/FCP-INDI/C-PAC/blob/80424468c7f4e59c102f446b05d4154ec1cd4b2d/CPAC/pipeline/engine.py#L1510-L1561) that makes the logic decision about where to pull.

Right now, it's too rigid with the assumed naming and path string formation. When providing a C-PAC output directory like this to the source directory field:

`/home/cpac_runs/postfreesurfer/postfs-output/output`

the function will assume the data directory is this:

`/home/cpac_runs/postfreesurfer/postfs-output/output/sub-0050952_ses-1/anat`

when it's really this:

`/home/cpac_runs/postfreesurfer/postfs-output/output/cpac_cpac_abcd-options/sub-0050952_ses-1/anat`

It will say nothing and proceed, it will find no outputs in the non-existing directory and continue with an empty resource pool.

In addition, this is overly C-PAC specific. If we want to make it easier to ingress output results from other pipelines, we will have to go towards something far more flexible.

This issue also involves the question about updating our `participant` and `session` directory levels via BIDS.

### Expected behavior

We need this feature to be more flexible, and to communicate more clearly to users when the paths misalign.

This doesn't involve BIDS parsing and the discussion of moving to PyBIDS, it's only the root path expectation and formation, and whether or not we want to expand it to check for lower level directories and automatically make any decisions.

### Acceptance criteria

- [ ] If we are going to update the participant-session hierarchy in the directories, then this needs to be resolved first.
- [ ] Error message if the assumed path does not exist.
- [ ] Log message announcing which resources/data have been successfully pulled into the resource pool from the source or output directory.
- [ ] (within reason) Automatic/automagic bridging of gaps, or providing suggestions to the user.

**Comment by shnizzedy** [comment]
> we could adopt [pyBIDS](https://github.com/bids-standard/pybids)?

I [installed PyBIDS](https://github.com/FCP-INDI/C-PAC/commit/be22ce92a458b1034ebbadfd84830400711d3a73) to use https://github.com/FCP-INDI/C-PAC/blob/be22ce92a458b1034ebbadfd84830400711d3a73/CPAC/qc/xcp.py#L383-L386 https://github.com/FCP-INDI/C-PAC/blob/be22ce92a458b1034ebbadfd84830400711d3a73/CPAC/qc/xcp.py#L351-L353 [for XCP QC](https://github.com/FCP-INDI/C-PAC/pull/1683) in 1.8.4 to avoid reinventing the wheel there, figured we'll phase it in throughout the rest of the code

**Comment by sgiavasis** [comment]
> @sgiavasis do you think that on the route to fixing this, we could adopt [pyBIDS](https://github.com/bids-standard/pybids)? This would also involve making our output directories more properly BIDS-compatible (e.g., separating session and subject).
> 
> At the hackathon next week I'll also be discussing [bids-standard/pybids#831](https://github.com/bids-standard/pybids/issues/831), which could be excellent for performance.

Yes, this is very high priority. We don't need PyBIDS to resolve this particular problem but we can fold it into the larger "data ingress upgrade" project.
